example_notebooks/cube_4d.ipynb

```diff
@@ -0,0 +1,354 @@
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Generate a fake 4D data cube."
+   ]
+  },
+  {
+   "cell_type": "raw",
+   "metadata": {},
+   "source": [
+    "import numpy as np\n",
+    "\n",
+    "pf = None  # Not using FITS file pointer\n",
+    "cube4d = np.random.random((2, 5, 100, 200))\n",
+    "\n",
+    "# Convert to NDData.\n",
+    "im = NDData(cube4d, wcs=WCS(naxis=4))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**--- OR ---**\n",
+    "\n",
+    "Load an existing data cube.\n",
+    "\n",
+    "Dask support requires https://github.com/ejeschke/ginga/pull/805."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import dask.array as da\n",
+    "from astropy.io import fits\n",
+    "\n",
+    "# This is a large cube that is 6GB.\n",
+    "filename = '/redkeep/ironthrone/ssb/stginga/test_data/NRCN815A_LIN_20160115_uncal.fits'\n",
+    "pf = fits.open(filename, memmap=True)\n",
+    "\n",
+    "# There is no WCS assigned for UNCAL file yet.\n",
+    "# This is a dummy WCS.\n",
+    "im_wcs = WCS(pf[1].header)\n",
+    "\n",
+    "# Convert to NDData that wraps a Dask array.\n",
+    "# Must provide Dask with a name to avoid hashing.\n",
+    "# Shape is (1, 190, 2048, 2048)\n",
+    "im = NDData(da.from_array(pf[1].data, name='nircam_sci'), wcs=im_wcs)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Grab the `z`-dimension for slider."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "n_3d = im.data.shape[1]"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Load `NDData` into viewer widget."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Loading a fake WCS into widget sometimes gives error.\n",
+    "# Try reloading before giving up.\n",
+    "w.load_nddata(im)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Create a slider for third dimension."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def show_slice(n):\n",
+    "    w.show_slice(n - 1)\n",
+    "    w.plot_line_profile()  # Update the red dot\n",
+    "\n",
+    "slider = ipyw.interactive(\n",
+    "    show_slice,\n",
+    "    n=ipyw.IntSlider(min=1, max=n_3d, step=1, value=0, continuous_update=False))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Set up the plot and hook to widget."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# If plot shows, rerun cell to hide it.\n",
+    "ax = plt.gca();\n",
```

Comment:
@mwcraig , do you know any notebook tricks to not show the Matplotlib window in this cell? It ways show on first run, but then disappears in subsequent runs in the same session.

Reply:
Hi, I think this works:
plt.ioff()

I've been using a version of this code in a notebook I've made for displaying HETDEX data cubes. I had to switch the axis around.

https://github.com/HETDEX/hetdex_api/blob/master/hetdex_api/cube_widget.py

Reply:
The plot doesn't seem to show at the wrong place for me anymore either; perhaps something was fixed upstream. Thanks for letting us know!

Switching axis is probably normal since images taken with different instruments/telescopes have different formats.